modyn/supervisor/internal/triggers/ensembletrigger.py

```diff
@@ -0,0 +1,230 @@
+from modyn.supervisor.internal.triggers.timetrigger import TimeTrigger
+from modyn.supervisor.internal.triggers.trigger import Trigger, TriggerContext
+from modyn.supervisor.internal.triggers.utils.models import (
+    EnsembleTriggerEvalLog,
+    TriggerPolicyEvaluationLog,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class EnsembleTrigger(Trigger):
+    """Evaluates multiple subtriggers and triggers if a certain after an
+    aggregation function aggregates subtrigger decisions to True.
+
+    Idea:
+        The EnsembleTrigger always evaluates and caches the next triggering index
+        from the generators of each subtrigger. The evaluations of all subsequent subtrigger indexes are
+        deferred in a lazy manner.
+
+        We assume that after a subtrigger will always like to see a trigger after the next index in the generator.
+        Thus, we are sure that we can trigger for certain at the maximum of all next trigger indexes of the subtriggers
+        as all subtriggers requested a trigger until this index.
+
+        For the potential triggering indexes up until this maximum, we evaluate the aggregation function at every
+        index where one policy first decided to do a trigger. We have those indexes cached after fetching the next
+        trigger index from the subtriggers.
+
+        We have to ensure though that every trigger consumes / evaluates all data eventually as some triggers
+        need the data to maintain their state.
+        Therefore we have to ensure right after a trigger, that all next trigger indexes in the cache are
+        in the future wr.t. our current trigger index. We can simply do that by fetching the next trigger indexes
+        from the subtriggers until all subtriggers have a next trigger index in the future.
+    """
+
+    def __init__(self, config: EnsembleTriggerConfig):
+        self.config = config
+        self.subtriggers = self._create_subtriggers(config)
+
+        self.last_inform_decisions: dict[str, bool] = {}
+        """Leftover decisions from the last inform call of the subtriggers that
+        didn't yield a trigger."""
+
+    def init_trigger(self, context: TriggerContext) -> None:
+        for trigger in self.subtriggers.values():
+            trigger.init_trigger(context)
+
+    def inform(
+        self,
+        new_data: list[tuple[int, int, int]],
+        log: TriggerPolicyEvaluationLog | None = None,
+    ) -> Generator[int, None, None]:
+        subtrigger_generators = {
+            trigger_name: trigger.inform(new_data, log) for trigger_name, trigger in self.subtriggers.items()
+        }
+
+        # Indicates whether the ensemble trigger has passed a triggering index
+        # with processing_head for a subtrigger since the last trigger.
+        # Will be reset after every trigger.
+        subtrigger_decision_cache = dict(
+            {trigger_name: False for trigger_name in self.subtriggers},
+            **self.last_inform_decisions,
+        )
```

Comment:
so this is basically a defaultdict with False, and you merge last_inform_decisions into that defaultdict? not sure whether using a defaultdict explicitly here will make the code more compact or not